### DISCUSSION TITLE
Displaying a particular variation sequentially out of four different variations of an experiment.

### DISCUSSION THREAD

**Discussion by united-forever** [jspsych/jsPsych#708] [Q&A]
Hi everyone,

I have four different variations of an experiment, and this is how I display a variation at random:

```js
var allVariations = ["/exp.html","/exp2.html","/exp3.html","/exp4.html"];
var randomizeVariations = allVariations[Math.floor(Math.random()*allVariations.length)];

document.getElementById('loadSurvey').onclick = function() {
	if (document.getElementById('consent_checkbox').checked) {
		window.location =  randomizeVariations;
	} else {
		alert('You need to check the checkbox to continue.');
		return false;
	}
	return false;
}
```

However , if  I wanted to display variations in a particular sequence i.e respondent 1 would see  variation2, respondent 2 would see variation1, respondent 3 would see variation4, respondent 4 would see variation3, and then the same pattern would repeat for successive respondents, how would I achieve that? I have a tried a no. of things but none seem to work.

Thanks,
Yash.

**Comment by jodeleeuw** [comment]
Anytime you need coordination across participants you are going to need some kind of server-side script. 

I've written a set of `PHP` scripts for my own lab that handle data storage and participant assignment. You can find them here: https://github.com/vassar-cogscilab/php-mysql-save-data.

There's a set of instructions in the readme file. These do require a MySQL database. You can do assignment like you are describing using the method described in step 7 of the readme. This involves registering each subject to a database table and computing the next condition on the server using PHP. The condition assignment is sent back to the experiment via an `XMLHttpRequest` object.

**Reply by united-forever** [reply]
Hey Josh,

Thanks for the prompt response! 

I get what you are suggesting, but I am using Node.js for my backend, and MongoDB for the storage.
Is there a way to code it for Node instead of PHP? I am fairly new to programming, and am coding an experiment, so grasping some of the basic concepts while coding is a little tough. :)

Thanks,
Yash.

**Reply by jodeleeuw** [reply]
It's definitely possible to code something like this in Node. Unfortunately I don't have any handy examples to share with you of how to do it in Node. If you've started to figure out how to handle data saving in Node, this might be a good next step for expanding your back end programming skills :smile:

**Reply by united-forever** [reply]
From what I could understand, the idea is to serve static html based on a conditional parameter generated in the backend(in my case I think it would something like 'if variation1 is done, then move on to the next one in the order'), and this data is dynamically communicated to the frontend.

Since I sort of have a deadline for this coming up, like tomorrow, it looks like the random display order will have to do for now. But it sounds like an interesting puzzle, so I will definitely look into how to get it done. Thanks for your input! If you have any other suggestions, I would definitely appreciate them!

Thanks,
Yash.

**Comment by vijaymarupudi** [comment]
I think a simple solution to this problem would be to just use links with different query parameters, and then assessing them at the beginning of the script. Unless you need the url to be the same everytime.

**Reply by jodeleeuw** [reply]
This is a great idea if you can manually send the link to each participant.

**Reply by united-forever** [reply]
Hey Vijay,

This seems like a reasonable solution. Could you provide an example of how I would actually assess them?

Thanks,
Yash.

**Reply by vijaymarupudi** [reply]
If you send participants a url like `https://url.com/experiment.html?experimentVersion=1`, then you can access it using

```javascript
jsPsych.data.getURLVariable('experimentVersion')
```

I will say that you might want to change the name of the key to something less obvious, or you'll might get participants tampering with the url.

**Reply by united-forever** [reply]
I had another idea. Say  I randomize display of the 4 variations. If a particular variation gets displayed a certain no. of times(a percentage of the total sample), then stop displaying that variation. Would it be possible to code this without back end validation?

**Reply by united-forever** [reply]
@vijaygopal1234  I am loading the four variations outside of jsPsych. How would I access the query paramters using `jsPsych.data.getURLVariable `in that case? I realize that one can use the external-html-plugin to load html files and execute JS code, but for some reason, that plugin is not working for me. I posted a query on the jsPsych google group page regarding this a while back.

**Reply by vijaymarupudi** [reply]
> I had another idea. Say I randomize display of the 4 variations. If a particular variation gets displayed a certain no. of times(a percentage of the total sample), then stop displaying that variation. Would it be possible to code this without back end validation?

No, not possible because it involves holding state beyond one trial. That requires backend memory.

>  I am loading the four variations outside of jsPsych. How would I access the query paramters using jsPsych.data.getURLVariable in that case? I realize that one can use the external-html-plugin to load html files and execute JS code, but for some reason, that plugin is not working for me. I posted a query on the jsPsych google group page regarding this a while back.

You can use https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams instead then.
